### PR TITLE
preview_database_id not being picked up

### PR DESCRIPTION
**Pull request by rozenmd**
Closes https://github.com/cloudflare/wrangler2/issues/2263

### CODE REVIEW

File: packages/wrangler/src/dev.tsx [hunk 1]
```diff
@@ -895,7 +895,17 @@ async function getBindings(
 		unsafe: configParam.unsafe?.bindings,
 		logfwdr: configParam.logfwdr,
 		d1_databases: identifyD1BindingsAsBeta([
-			...configParam.d1_databases,
+			...(configParam.d1_databases || []).map((d1Db) => {
+				if (!d1Db.preview_database_id) {
+					throw new Error(
```
Comment (GregBrimble) [inline packages/wrangler/src/dev.tsx:900]: IIRC, KV warns but doesn't error. Do you want the same behavior here?

**Review by GregBrimble — APPROVED** [review]

File: packages/wrangler/src/dev.tsx (hunk 1 above)
Comment (rozenmd) [reply packages/wrangler/src/dev.tsx:900]: nah it throws
image

File: packages/wrangler/src/config/index.ts [hunk 2]
```diff
@@ -159,14 +159,21 @@ export function printBindings(bindings: CfWorkerInit["bindings"]) {
...
+				({ binding, database_name, database_id, preview_database_id }) => {
+					let databaseValue = `${database_id}`;
+					if (database_name) {
+						databaseValue = `${database_name} (${database_id})`;
+					}
+					if (preview_database_id) {
+						databaseValue += `, Preview: (${preview_database_id})`;
+					}
+					return {
+						key: removeD1BetaPrefix(binding),
```
Comment (JacobMGEvans) [inline packages/wrangler/src/config/index.ts:172]: Any tests to handle the control logic or confirm the bug is not reproduced?

**Review by JacobMGEvans — APPROVED** [review]

File: packages/wrangler/src/config/index.ts (hunk 2 above)
Comment (rozenmd) [reply packages/wrangler/src/config/index.ts:172]: That bit isn't the bug, I just wanted to print what we had access to

File: packages/wrangler/src/dev.tsx [hunk 3]
```diff
@@ -895,7 +895,17 @@ async function getBindings(
...
+						`In development, you should use a separate d1 database than the one you'd use in production. Please create a new d1 database with "wrangler d1 create <name>" and add its id as preview_database_id to the d1_database "${d1Db.binding}" in your wrangler.toml`
+					);
+				}
+				return {
+					...d1Db,
+					database_id: d1Db.preview_database_id,
```
Comment (rozenmd) [inline packages/wrangler/src/dev.tsx:906]: bug is fixed here @JacobMGEvans - but it throws when I try it out

File: packages/wrangler/src/dev.tsx [hunk 4]
```diff
@@ -895,7 +895,20 @@ async function getBindings(
 		unsafe: configParam.unsafe?.bindings,
 		logfwdr: configParam.logfwdr,
 		d1_databases: identifyD1BindingsAsBeta([
-			...configParam.d1_databases,
+			...(configParam.d1_databases || []).map((d1Db) => {
+				if (!d1Db.preview_database_id) {
+					throw new Error(
+						`In development, you should use a separate d1 database than the one you'd use in production. Please create a new d1 database with "wrangler d1 create <name>" and add its id as preview_database_id to the d1_database "${d1Db.binding}" in your wrangler.toml`
```
Comment (penalosa) [inline packages/wrangler/src/dev.tsx:901]:
```suggestion
						`In development, you should use a separate D1 database than the one you'd use in production. Please create a new D1 database with "wrangler d1 create <name>" and add its id as preview_database_id to the d1_database "${d1Db.binding}" in your wrangler.toml`
```

File: packages/wrangler/src/dev.tsx [hunk 5]
```diff
@@ -895,7 +895,20 @@ async function getBindings(
...
+					);
+				}
+				const newD1Db = {
+					...d1Db,
+					database_id: d1Db.preview_database_id,
+				};
+				console.log("old: ", d1Db);
+				console.log("new: ", newD1Db);
```
Comment (penalosa) [inline packages/wrangler/src/dev.tsx:909]: Should the logger be used for this? (probably `logger.debug`)?

File: packages/wrangler/src/dev.tsx [hunk 6]
```diff
@@ -895,7 +895,20 @@ async function getBindings(
...
 		logfwdr: configParam.logfwdr,
 		d1_databases: identifyD1BindingsAsBeta([
-			...configParam.d1_databases,
+			...(configParam.d1_databases || []).map((d1Db) => {
+				if (!d1Db.preview_database_id) {
+					throw new Error(
+						`In development, you should use a separate d1 database than the one you'd use in production. Please create a new d1 database with "wrangler d1 create <name>" and add its id as preview_database_id to the d1_database "${d1Db.binding}" in your wrangler.toml`
+					);
+				}
+				const newD1Db = {
```
Comment (penalosa) [inline packages/wrangler/src/dev.tsx:904]: Can this have a more descriptive name? Maybe previewD1Db or something?

**Comment by rozenmd** [comment]
Just moved this PR back to draft @JacobMGEvans @penalosa - still trying to figure out what's going on

**Comment by rozenmd** [comment]
disregard that @penalosa @JacobMGEvans - skill issue. I was typoing the preview ID. It works.

**Pull request merged**